linker.ts

```diff
@@ -1,72 +1,132 @@
+ */
+function replacePlaceholderLabelWithHexAddress (bytecode, hexAddress, labelName) {
+  // truncate to 36 characters, why, I assume this is because legacy names
+  // only had a length of 36 characters?
+  const truncatedName = labelName.slice(0, 36);
+
+  const prefix = '__';
+  const postFix = `${Array(37 - truncatedName.length).join('_')}__`;
+  const libLabel = `${prefix}${truncatedName}${postFix}`;
+
+  while (bytecode.indexOf(libLabel) >= 0) {
+    bytecode = bytecode.replace(libLabel, hexAddress);
+  }
+
+  return bytecode;
+}
+
+/**
+ * When using libraries, the resulting bytecode will contain placeholders for
+ * the real addresses of the referenced libraries. These have to be updated,
+ * via a process called linking, before deploying the contract.
+ *
+ * This method provides a simple helper for linking.
+ *
+ * @param {string} bytecode
+ * @param {string} libraries
+ * @returns {string} bytecode
+ */
+function linkBytecode (bytecode, libraries) {
   assert(typeof bytecode === 'string');
   assert(typeof libraries === 'object');
+
   // NOTE: for backwards compatibility support old compiler which didn't use file names
   const librariesComplete = {};
-  for (const libraryName in libraries) {
-    if (typeof libraries[libraryName] === 'object') {
-      // API compatible with the standard JSON i/o
-      for (const lib in libraries[libraryName]) {
-        librariesComplete[lib] = libraries[libraryName][lib];
-        librariesComplete[libraryName + ':' + lib] = libraries[libraryName][lib];
-      }
-    } else {
-      // backwards compatible API for early solc-js versions
-      const parsed = libraryName.match(/^([^:]+):(.+)$/);
-      if (parsed) {
-        librariesComplete[parsed[2]] = libraries[libraryName];
+
+  for (const [libraryName, library] of Object.entries(libraries)) {
+    if (isNil(library)) continue;
+
+    // API compatible with the standard JSON i/o
+    if (isObject(library)) {
+      for (const [key, value] of Object.entries(library)) {
+        librariesComplete[key] = value;
+        librariesComplete[`${libraryName}:${key}`] = value;
       }
-      librariesComplete[libraryName] = libraries[libraryName];
+
+      continue;
     }
+
+    // backwards compatible API for early solc-js versions
+    const parsed = libraryName.match(/^([^:]+):(.+)$/);
+
+    if (!isNil(parsed)) librariesComplete[parsed[2]] = library;
+    librariesComplete[libraryName] = library;
   }
 
   for (const libraryName in librariesComplete) {
     let hexAddress = librariesComplete[libraryName];
-    if (hexAddress.slice(0, 2) !== '0x' || hexAddress.length > 42) {
-      throw new Error('Invalid address specified for ' + libraryName);
+
+    if (!hexAddress.startsWith('0x') || hexAddress.length > 42) {
+      throw new Error(`Invalid address specified for ${libraryName}`);
     }
+
     // remove 0x prefix
     hexAddress = hexAddress.slice(2);
     hexAddress = Array(40 - hexAddress.length + 1).join('0') + hexAddress;
```

Comment:
I wanted to suggest creating a helper for padding but looks like one exists already in `String`:
```suggestion
    hexAddress = hexAddress.slice(2).padStart('0', 40);
```